### ISSUE TITLE
Setup configure's test if library is available can spuriously fail in certain circumstances

### ISSUE THREAD

**Issue by ezyang** [haskell/cabal#4544]
To test if a library libfoo is installed, we try building a simple program with `gcc test.c -lfoo`. Ordinarily this will work even if foo depends on other libraries, because gcc will either (a) fail because foo doesn't exist, or (b) succeed, because foo exist, and ld is too lazy to actually try linking foo if none of the symbols are used.

However, if you pass `-Wl,-no-as-needed`, then gcc will eagerly link the library, and now you are in trouble, because this will fail unless you also pass all the other libraries you need. That's not right.

We should check autogoo to see what a robust way of testing if a library exists is.

I repro'd this from #4541, but #4541 is asking for something different (it is a meta-issue asking us to stop trying to be so clever here.)

**Comment by hvr** [comment]
...what is "autogoo"? I hope you're not trying to make fun of GNU Autoconf! ;-)

**Comment by suzumiyasmith** [comment]
Since there is not a clear solution now, maybe we could change this warning message from "lib missing" to "lib missing or lib broken".

**Comment by ezyang** [comment]
@suzumiyasmith Yes, would be very happy to accept a simpler patch that did this. Interested? I can advise :)
